lib/barbeque/executor/docker.rb

```diff
@@ -1,35 +1,125 @@
 require 'barbeque/docker_image'
+require 'barbeque/slack_notifier'
 require 'open3'
 
 module Barbeque
   module Executor
     class Docker
+      class DockerCommandError < StandardError
+      end
+
       def initialize(_options)
       end
 
       # @param [Barbeque::JobExecution] job_execution
       # @param [Hash] envs
-      # @return [String] stdout
-      # @return [String] stderr
-      # @return [Process::Status] status
-      def run(job_execution, envs)
-        job_definition = job_execution.job_definition
-        docker_image = DockerImage.new(job_definition.app.docker_image)
-        cmd = build_docker_run_command(docker_image, command, envs)
-        Bundler.with_clean_env { Open3.capture3(*cmd) }
+      def start_execution(job_execution, envs)
+        docker_image = DockerImage.new(job_execution.job_definition.app.docker_image)
+        cmd = build_docker_run_command(docker_image, job_execution.job_definition.command, envs)
+        stdout, stderr, status = Open3.capture3(*cmd)
+        if status.success?
+          job_execution.update!(status: :running)
+          Barbeque::DockerContainer.create!(message_id: job_execution.message_id, container_id: stdout.chomp)
+        else
+          job_execution.update!(status: :failed, finished_at: Time.zone.now)
+          Barbeque::ExecutionLog.save_stdout_and_stderr(job_execution, stdout, stderr)
+          Barbeque::SlackNotifier.notify_job_execution(job_execution)
+        end
+      end
+
+      # @param [Barbeque::JobRetry] job_retry
+      # @param [Hash] envs
+      def start_retry(job_retry, envs)
+        job_execution = job_retry.job_execution
+        docker_image = DockerImage.new(job_execution.job_definition.app.docker_image)
+        cmd = build_docker_run_command(docker_image, job_execution.job_definition.command, envs)
+        stdout, stderr, status = Open3.capture3(*cmd)
+        if status.success?
+          job_execution.update!(status: :retried)
+          job_retry.update!(status: :running)
```

Comment:
To avoid having unexpected data in the case barbeque-worker unexpectedly shuts down immediately before this line, L38-39 may need to be applied in transaction.